homeassistant/components/device_tracker/icloud.py

```diff
@@ -47,39 +50,39 @@ def keep_alive(now):
         api.authenticate()
         _LOGGER.info("Authenticate against iCloud")
 
-    track_utc_time_change(hass, keep_alive, second=0)
-
     def update_icloud(now):
         """Authenticate against iCloud and scan for devices."""
         try:
             # The session timeouts if we are not using it so we
             # have to re-authenticate. This will send an email.
-            api.authenticate()
+            keep_alive(None)
             # Loop through every device registered with the iCloud account
             for device in api.devices:
                 status = device.status()
                 location = device.location()
                 # If the device has a location add it. If not do nothing
                 if location:
                     see(
-                        dev_id=re.sub(r"(\s|\W|')",
-                                      '',
-                                      status['name']),
+                        dev_id=slugify(status['name'].replace(' ', '', 99)),
                         host_name=status['name'],
                         gps=(location['latitude'], location['longitude']),
                         battery=status['batteryLevel']*100,
                         gps_accuracy=location['horizontalAccuracy']
                     )
-                else:
-                    # No location found for the device so continue
-                    continue
         except PyiCloudNoDevicesException:
             _LOGGER.info('No iCloud Devices found!')
 
-    track_utc_time_change(
-        hass, update_icloud,
-        minute=range(0, 60, config.get(CONF_INTERVAL, DEFAULT_INTERVAL)),
-        second=0
-    )
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, update_icloud)
+
+    interval = config[CONF_INTERVAL]
+    track_utc_time_change(hass, update_icloud, second=0,
+                          minute=range(0, 60, interval))
+
+    # Schedule keepalives between the updates
+    keep_alive_start = KEEPALIVE_INTERVAL
```

Comment:
I changed it because I hated the fixed interval. If the interval was too high, the update was too late and if the interval was small, the battery would be drained too much. You should check my repo HA_icloud.
The logic I used is the following: when in a known zone the interval is 30 minutes. Otherwise the interval depends on the distance towards home and the battery level. This will make sure you get seen at home as soon as you arrive and if you're far away the battery isn't stressed. If you're on vacation the interval would even increase more.
The interval in my component is also device based. If my ipad is at home but I'm away, I don't need my ipad to be drained also. I also give the option to ignore some devices so they don't get checked all the time.

Op 25 jul. 2016 om 23:17 heeft Johann Kellerman <notifications@github.com<mailto:notifications@github.com>> het volgende geschreven:


In homeassistant/components/device_tracker/icloud.py<https://github.com/home-assistant/home-assistant/pull/2614#discussion_r72146647>:

>          except PyiCloudNoDevicesException:
>              _LOGGER.info<http://LOGGER.info>('No iCloud Devices found!')
>
> -    track_utc_time_change(
> -        hass, update_icloud,
> -        minute=range(0, 60, config.get(CONF_INTERVAL, DEFAULT_INTERVAL)),
> -        second=0
> -    )
> +    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, update_icloud)
> +
> +    interval = config[CONF_INTERVAL]
> +    track_utc_time_change(hass, update_icloud, second=0,
> +                          minute=range(0, 60, interval))
> +
> +    # Schedule keepalives between the updates
> +    keep_alive_start = KEEPALIVE_INTERVAL


@Bart274<https://github.com/Bart274> interesting. It would be really great if the frontend can show these properties!

What is the main reason for your intervals? Don't the iphone update icloud regardless if its read or not?

I'm pretty sure you can have multiple accounts today. Used it for a short while until I discovered Family accounts shows all our devices with a single login (they cant all be called "iphone" though ?)

device_tracker:
  - platform: icloud
    ...
  - platform: icloud
    ...

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/2614/files/99e5954b9b4530b69c838cfc0617c9895162c7ec#r72146647>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABERpglhUnkAiSg83YfD-F1YrkwmPq1Pks5qZSfVgaJpZM4JTmyj>.